[MySensors] No MQTT connection with OH2.5

<!-- Provide a general summary of the issue in the *Title* above -->
<!-- If the issue is related to a binding, please include its short name in -->
<!-- square brackets in the title - Example: "[astro] My issue..." -->

<!-- Important: Please contact the openHAB community forum for questions or -->
<!-- for configuration and usage guidance: https://community.openhab.org -->

<!-- Feel free to delete any comment lines in the template (starting with "<!--") -->

## Expected Behavior
After starting OH or adding the MySensors binding jar to addon folder a MQTT connection to the MQTT broker should be establiehed.

## Current Behavior
<!-- If describing a bug, tell us what happens instead of the expected behavior -->
<!-- Include related log information (preferably debug level) and related configs. -->
<!-- Use file attachment for log and config information longer than a few lines -->
<!-- If suggesting a change/improvement, explain the difference from current behavior -->
<!-- For improvements, discuss at community.openhab.org first and include link to topic -->
With OH 2.5 and MySensors binding as osgi bundle its not possible to get a working connection to MQTT broker. In the log you get:
 ```
[ERROR] [rotocol.mqtt.MySensorsMqttConnection] - MqttService is null!
[ERROR] [rotocol.mqtt.MySensorsMqttConnection] - Failed connecting to bridge...next retry in 10 seconds (Retry No.:0)
```
No connection could be establiehed.

## Possible Solution
See [https://community.openhab.org/t/oh-2-5-snapshot-and-mysensors-binding/86580/29](url).
Until clarified why the osgi dependency injection for MqttService does not work we can use 
a workaround in MySensorsMqttConnection.establishConnection():
```
       boolean connectionEstablished = false;

        //Start workaround
        if (MySensorsMqttService.getMqttService() == null) {
            BundleContext bc = FrameworkUtil.getBundle(getClass()).getBundleContext();
            final ServiceReference sr = bc.getServiceReference(MqttService.class.getName());
            if (sr != null) {
                final MqttService mqtt = (MqttService) bc.getService(sr);
                if (mqtt != null) {
                    new MySensorsMqttService().setMqttService(mqtt);
                }
            }
        }
        //End workaround

        if (MySensorsMqttService.getMqttService() == null) {
```
Shortly after starting with this binding you see in the log:
```
        [INFO ] [rotocol.mqtt.MySensorsMqttConnection] - Successfully connected to MySensors Bridge.
```
and the things are going online in paperui.
## Steps to Reproduce (for Bugs)
<!-- Provide a link to a live example, or an unambiguous set of steps to -->
<!-- reproduce this bug. Include code to reproduce, if relevant -->
1. Have a system broker with name "mosquitto" defined via file "org.eclipse.smarthome.mqttbroker.cfg"
2. Define a MySensors MQTT Bridge via mysensors.things file
```
Bridge mysensors:bridge-mqtt:gwMQ01 [   brokerName="mosquitto", 
                                        topicPublish="mygw1-in",
                                        topicSubscribe="mygw1-out",
                                        startupCheckEnabled=false ] {
										
    // no things yet
}
```
3. Use development branch "MySensors_Binding_bnd" and enable MQTT again in MySensorsBridgeHandler.java:
```
        } else if (bridgeuid.equals(THING_TYPE_BRIDGE_MQTT)) {
            gatewayConfig.setGatewayType(MySensorsGatewayType.MQTT);
            gatewayConfig.setBrokerName(conf.brokerName);
```
4. Build org.openhab.binding.mysensors-2.5.0-SNAPSHOT.jar and add it to addons folder
5. Clear (delete contents of) userdata/cache folder to abandon old versions of binding
6. Start OH2.5
7. Add missing dependencies via karaf console:
```
feature:install openhab-core-io-transport-mqtt
feature:install openhab-transport-serial
```
8. See errors in log (see above) or in paperui the 

## Context
Since changing to osgi bundling of the MySensors binding the dependency injection of the MQTT service ist not working.

## Your Environment
* OH 2.5 and MySensors binding 2.5 (org.openhab.binding.mysensors-2.5.0-SNAPSHOT.jar from above)
* Java 8
* Windows 10 standalone or with Docker image "openhab/openhab:2.5.0-amd64-debian" on Linux

--- a/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/factory/MySensorsCacheFactory.java
+++ b/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/factory/MySensorsCacheFactory.java
@@ -75,7 +75,7 @@ public class MySensorsCacheFactory {
         T ret = def;
 
         try {
-            File f = new File(CACHE_BASE_PATH + "/" + GIVEN_IDS_CACHE_FILE + CACHE_FILE_SUFFIX);
+            File f = new File(CACHE_BASE_PATH + "/" + fileName + CACHE_FILE_SUFFIX);// FTi
 
             if (f.exists()) {
                 logger.debug("Cache file: {} exist.", GIVEN_IDS_CACHE_FILE + CACHE_FILE_SUFFIX);
@@ -106,7 +106,7 @@ public class MySensorsCacheFactory {
     private synchronized <T> void jsonToFile(String fileName, T obj, Type clasz) {
         JsonWriter jsonWriter = null;
         try {
-            File f = new File(CACHE_BASE_PATH + "/" + GIVEN_IDS_CACHE_FILE + CACHE_FILE_SUFFIX);
+            File f = new File(CACHE_BASE_PATH + "/" + fileName + CACHE_FILE_SUFFIX);// FTi
 
             jsonWriter = new JsonWriter(new FileWriter(f));
 

--- a/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/handler/MySensorsBridgeHandler.java
+++ b/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/handler/MySensorsBridgeHandler.java
@@ -156,15 +156,19 @@ public class MySensorsBridgeHandler extends BaseBridgeHandler implements MySenso
     private void updateCacheFile() {
         List<Integer> givenIds = myGateway.getGivenIds();
 
-        cacheFactory.writeCache(MySensorsCacheFactory.GIVEN_IDS_CACHE_FILE, givenIds.toArray(new Integer[] {}),
-                Integer[].class);
+        String cacheFileName = MySensorsCacheFactory.GIVEN_IDS_CACHE_FILE + "_"
+                + getThing().getUID().toString().replace(':', '_');
+
+        cacheFactory.writeCache(cacheFileName, givenIds.toArray(new Integer[] {}), Integer[].class);
     }
 
     private Map<Integer, MySensorsNode> loadCacheFile() {
         Map<Integer, MySensorsNode> nodes = new HashMap<Integer, MySensorsNode>();
 
-        List<Integer> givenIds = cacheFactory.readCache(MySensorsCacheFactory.GIVEN_IDS_CACHE_FILE,
-                new ArrayList<Integer>(), new TypeToken<ArrayList<Integer>>() {
+        String cacheFileName = MySensorsCacheFactory.GIVEN_IDS_CACHE_FILE + "_"
+                + getThing().getUID().toString().replace(':', '_');
+        List<Integer> givenIds = cacheFactory.readCache(cacheFileName, new ArrayList<Integer>(),
+                new TypeToken<ArrayList<Integer>>() {
                 }.getType());
 
         for (Integer i : givenIds) {

--- a/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/handler/MySensorsBridgeHandler.java
+++ b/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/handler/MySensorsBridgeHandler.java
@@ -190,6 +190,7 @@ public class MySensorsBridgeHandler extends BaseBridgeHandler implements MySenso
             gatewayConfig.setIpAddress(conf.ipAddress);
             gatewayConfig.setTcpPort(conf.tcpPort);
         } else if (bridgeuid.equals(THING_TYPE_BRIDGE_MQTT)) {
+            gatewayConfig.setGatewayType(MySensorsGatewayType.MQTT);
             gatewayConfig.setBrokerName(conf.brokerName);
             gatewayConfig.setTopicPublish(conf.topicPublish);
             gatewayConfig.setTopicSubscribe(conf.topicSubscribe);

--- a/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/mqtt/MySensorsMqttConnection.java
+++ b/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/mqtt/MySensorsMqttConnection.java
@@ -45,8 +45,8 @@ public class MySensorsMqttConnection extends MySensorsAbstractConnection impleme
     private MySensorsMqttSubscriber myMqttSub;
     private MqttBrokerConnection connection;
 
-    private static PipedOutputStream out;
-    private static PipedInputStream in;
+    private PipedOutputStream out;
+    private PipedInputStream in;
 
     private MySensorsMqttPublishCallback myMqttPublishCallback = new MySensorsMqttPublishCallback();
 

--- a/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/mqtt/MySensorsMqttConnection.java
+++ b/bundles/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/mqtt/MySensorsMqttConnection.java
@@ -23,10 +23,14 @@ import org.eclipse.smarthome.io.transport.mqtt.MqttBrokerConnection;
 import org.eclipse.smarthome.io.transport.mqtt.MqttConnectionObserver;
 import org.eclipse.smarthome.io.transport.mqtt.MqttConnectionState;
 import org.eclipse.smarthome.io.transport.mqtt.MqttMessageSubscriber;
+import org.eclipse.smarthome.io.transport.mqtt.MqttService;
 import org.openhab.binding.mysensors.internal.event.MySensorsEventRegister;
 import org.openhab.binding.mysensors.internal.gateway.MySensorsGatewayConfig;
 import org.openhab.binding.mysensors.internal.protocol.MySensorsAbstractConnection;
 import org.openhab.binding.mysensors.internal.protocol.message.MySensorsMessage;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
 
 /**
  * Implements the MQTT connection to a gateway of the MySensors network.
@@ -38,7 +42,6 @@ import org.openhab.binding.mysensors.internal.protocol.message.MySensorsMessage;
 
 public class MySensorsMqttConnection extends MySensorsAbstractConnection implements MqttConnectionObserver {
 
-    private MqttBrokerConnection mqttBrokerConn;
     private MySensorsMqttSubscriber myMqttSub;
     private MqttBrokerConnection connection;
 
@@ -58,6 +61,19 @@ public class MySensorsMqttConnection extends MySensorsAbstractConnection impleme
     @Override
     protected boolean establishConnection() {
         boolean connectionEstablished = false;
+
+        // ## Start workaround
+        if (MySensorsMqttService.getMqttService() == null) {
+            BundleContext bc = FrameworkUtil.getBundle(getClass()).getBundleContext();
+            final ServiceReference sr = bc.getServiceReference(MqttService.class.getName());
+            if (sr != null) {
+                final MqttService mqtt = (MqttService) bc.getService(sr);
+                if (mqtt != null) {
+                    new MySensorsMqttService().setMqttService(mqtt);
+                }
+            }
+        }
+        // ## End workaround
 
         if (MySensorsMqttService.getMqttService() == null) {
             logger.error("MqttService is null!");
@@ -105,7 +121,7 @@ public class MySensorsMqttConnection extends MySensorsAbstractConnection impleme
     protected void stopConnection() {
         in = null;
         out = null;
-        mqttBrokerConn.unsubscribe(myMqttSub.getTopic(), myMqttSub);
+        connection.unsubscribe(myMqttSub.getTopic(), myMqttSub);
     }
 
     /**
@@ -124,6 +140,11 @@ public class MySensorsMqttConnection extends MySensorsAbstractConnection impleme
 
         @Override
         public void processMessage(String topic, byte[] payload) {
+            if (!isConnected() || out == null) {
+                logger.error("Not connected - Messages can not be processed");
+                return;
+            }
+
             String payloadString = new String(payload);
             logger.debug("MQTT message received. Topic: {}, Message: {}", topic, payloadString);
             if (topic.indexOf(myGatewayConfig.getTopicSubscribe()) == 0) {

